src/main/kotlin/org/jitsi/nlj/rtp/codec/vp8/Vp8Packet.kt

```diff
@@ -17,48 +17,76 @@
 package org.jitsi.nlj.rtp.codec.vp8
 
 import org.jitsi.nlj.codec.vp8.Vp8Utils
-import org.jitsi.nlj.rtp.VideoRtpPacket
+import org.jitsi.nlj.rtp.ParsedVideoPacket
 import org.jitsi.nlj.util.cwarn
 import org.jitsi.nlj.util.getLogger
 import org.jitsi.rtp.extensions.bytearray.hashCodeOfSegment
 import org.jitsi_modified.impl.neomedia.codec.video.vp8.DePacketizer
 
-class Vp8Packet(
-    data: ByteArray,
+/**
+ * If this [Vp8Packet] instance is being created via a clone,
+ * we've already parsed the packet itself and determined whether
+ * or not its a keyframe and what its spatial layer index is,
+ * so the constructor allows passing in those values if
+ * they're already known.  If they're null, this instance
+ * will do the parsing itself
+ */
+class Vp8Packet private constructor (
+    buffer: ByteArray,
     offset: Int,
-    length: Int
-) : VideoRtpPacket(data, offset, length) {
+    length: Int,
+    isKeyframe: Boolean?,
+    isStartOfFrame: Boolean?,
+    spatialLayerIndex: Int?
+) : ParsedVideoPacket(buffer, offset, length) {
+
+    constructor(
+        buffer: ByteArray,
+        offset: Int,
+        length: Int
+    ) : this(buffer, offset, length,
+        isKeyframe = null,
+        isStartOfFrame = null,
+        spatialLayerIndex = null
+    )
+
+    override val isKeyframe: Boolean = isKeyframe ?: DePacketizer.isKeyFrame(this.buffer, offset, length)
+
+    val isStartOfFrame: Boolean = isStartOfFrame ?: DePacketizer.VP8PayloadDescriptor.isStartOfFrame(buffer, offset)
```

Comment:
should this be abstract in `ParsedVideoPacket`?

Reply:
I guess the question is whether I'll be using `ParsedVideoPacket` directly anywhere.

It's possible I may be able to write the new frame projection in terms of `ParsedVideoPacket` -- if so, that'd make it a lot easier to implement new codecs in the jvb. I'll need to consider how doable this will be, though.

Reply:
It turns out there's enough VP8-specific code in the VP8 frame projection logic that there isn't much point in making this generic at this point.

In the future if we add support for additional codecs it's probably worth investigating what can be shared and thus what should be abstract methods of `ParsedVideoPacket`.